app/src/main/java/org/thoughtcrime/securesms/payments/PaymentParcelable.java

```diff
@@ -0,0 +1,215 @@
+
+    if (payment.getFailureReason() == null) {
+      dest.writeInt(-1);
+    } else {
+      dest.writeInt(payment.getFailureReason().serialize());
+    }
+
+    dest.writeString(payment.getNote());
+    dest.writeString(payment.getAmount().serialize());
+    dest.writeString(payment.getFee().serialize());
+    dest.writeByteArray(payment.getPaymentMetaData().toByteArray());
+    dest.writeByte(payment.isSeen() ? (byte) 1 : 0);
+    dest.writeString(payment.getAmountWithDirection().serialize());
+    dest.writeString(payment.getAmountPlusFeeWithDirection().serialize());
+    dest.writeByte(payment.isDefrag() ? (byte) 1 : 0);
+  }
+
+  private static final class ParcelPayment implements Payment {
+
+    private final UUID            uuid;
+    private final Payee           payee;
+    private final long            blockIndex;
+    private final long            blockTimestamp;
+    private final long            timestamp;
+    private final long            displayTimestamp;
+    private final Direction       direction;
+    private final State           state;
+    private final FailureReason   failureReason;
+    private final String          note;
+    private final Money           amount;
+    private final Money           fee;
+    private final PaymentMetaData paymentMetaData;
+    private final boolean         isSeen;
+    private final Money           amountWithDirection;
+    private final Money           amountPlusFeeWithDirection;
+    private final boolean         isDefrag;
+
+    private ParcelPayment(Parcel in) {
+      try {
+        uuid = UUID.fromString(in.readString());
+
+        PayeeParcelable payeeParcelable = in.readParcelable(PayeeParcelable.class.getClassLoader());
+        payee = payeeParcelable.getPayee();
+
+        blockIndex                 = in.readLong();
+        blockTimestamp             = in.readLong();
+        timestamp                  = in.readLong();
+        displayTimestamp           = in.readLong();
+        direction                  = Direction.deserialize(in.readInt());
+        state                      = State.deserialize(in.readInt());
+
+        int failureReasonSerialized = in.readInt();
+        if (failureReasonSerialized == -1) {
+          failureReason = null;
+        } else {
+          failureReason = FailureReason.deserialize(failureReasonSerialized);
+        }
+
+        note                       = in.readString();
+        amount                     = Money.parse(in.readString());
+        fee                        = Money.parse(in.readString());
+        paymentMetaData            = PaymentMetaData.parseFrom(in.createByteArray());
+        isSeen                     = in.readByte() == 1;
+        amountWithDirection        = Money.parse(in.readString());
+        amountPlusFeeWithDirection = Money.parse(in.readString());
+        isDefrag                   = in.readByte() == 1;
+      } catch (Money.ParseException | InvalidProtocolBufferException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)